### PR TITLE
New Parser for Quark

### PR DESCRIPTION
**Pull request by rachitnigam**
Small changes to the BNF (adding parenthesis for functions) where needed.

### CODE REVIEW

File: js/src/main/scala/QuarkParser.scala [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    expr
+  )
+
+  lazy val seq: P[Expr] = positioned (
+    expr ~ ";" ~ seq ^^ { case s ~ _ ~ e => ESeq(s, e) } |
+    app
+  )
+}
+
+object QuarkParser {
```
Comment (arjunguha) [inline js/src/main/scala/QuarkParser.scala]: Yes! Object name matches file name! Java programmers rejoice!

File: js/src/test/scala/BrowserTestSuite.scala [hunk 2]
```diff
@@ -64,7 +64,7 @@ class BrowserTestSuite extends org.scalatest.AsyncFunSuite {
   }
 
   test("adding two numbers") {
-    parseAndEval("""add(1, 200)""")
+    parseAndEval("""op(1 + 200)""")
```
Comment (arjunguha) [inline js/src/test/scala/BrowserTestSuite.scala]: strange. Why `op(1 + 200)`. Why not just `1 + 200`?

File: js/src/test/scala/BrowserTestSuite.scala [hunk 3]
```diff
@@ -274,17 +265,16 @@ class BrowserTestSuite extends org.scalatest.AsyncFunSuite {
...
-        y := True
-      }else {
-        y := False
-      };
-      if(y){
-        "Pass < 50"
-      }else{
-        "Pass > 50"
-      }
+      let y = false in
```
Comment (arjunguha) [inline js/src/test/scala/BrowserTestSuite.scala]: I appreciate the lack of curly braces.

File: js/src/test/scala/ParserTests.scala [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    assert(parse("""-1""") == Const(CNum(-1.0)))
+  }
+
+  test("1. Testing multiple statements"){
+    val trial  = "8; true"
+    assert(parse(trial) == ESeq(Const(CNum(8)), Const(CBool(true))))
+  }
+
+  test("App"){
+    assert(parse("(1)(2)") == App(Const(CNum(1)), List(Const(CNum(2)))))
```
Comment (arjunguha) [inline js/src/test/scala/ParserTests.scala]: Are the parentheses still needed?

File: shared/src/main/scala/Syntax.scala [hunk 5]
```diff
@@ -19,15 +19,15 @@ object Syntax {
...
+  sealed trait Constant extends Positional
   case class CNum(n: Double) extends Constant
   case class CStr(s: String) extends Constant
   case class CBool(b: Boolean) extends Constant
   case class CRegex(s: String) extends Constant
   case class CUndefined() extends Constant
   case class CNull() extends Constant
 
-  sealed trait Expr extends WithPos {
+  sealed trait Expr extends WithPos with Positional {
```
Comment (arjunguha) [inline shared/src/main/scala/Syntax.scala]: Yikes. OK, so now we have two ways of setting positions on expressions. I'm not super thrilled about that. Why don't we create our own version of the `positioned` function that works with `WithPos`?

File: shared/src/main/scala/Syntax.scala [hunk 6]
```diff
@@ -19,15 +19,15 @@ object Syntax {
   type View = Set[Label]
   val emptyView: View = Set()
 
-  sealed trait Constant
+  sealed trait Constant extends Positional
```
Comment (arjunguha) [inline shared/src/main/scala/Syntax.scala]: Let's not bother with this, IMO.

File: shared/src/main/scala/Syntax.scala [hunk 7]
```diff
@@ -97,14 +97,14 @@ object Syntax {
...
-  case class Assign(x: String, e: Expr) extends Expr
+  case class Assign(x: Id, e: Expr) extends Expr
   case class Facet(k: Label, e1: Expr, e2: Expr) extends Expr
   case class Declassify(e: Expr) extends Expr
   case class Endorse(e: Expr) extends Expr
   case class ITE(pred: Expr, e1: Expr, e2: Expr) extends Expr
   case class While(pred: Expr, body: Expr) extends Expr
   case class Break(label: String, ret: Expr) extends Expr
-  case class ELabel(label: List[String], e: Expr) extends Expr
+  case class ELabel(label: List[Id], e: Expr) extends Expr
```
Comment (arjunguha) [inline shared/src/main/scala/Syntax.scala]: Good.

File: shared/src/main/scala/Syntax.scala [hunk 8]
```diff
@@ -97,14 +97,14 @@ object Syntax {
   case class Const(c: Constant) extends Expr
   case class Fun(x: List[Id], e: Expr) extends Expr
   case class App(e1: Expr, args: List[Expr]) extends Expr
-  case class Assign(x: String, e: Expr) extends Expr
+  case class Assign(x: Id, e: Expr) extends Expr
```
Comment (arjunguha) [inline shared/src/main/scala/Syntax.scala]: Good.

File: js/src/test/scala/ParserTests.scala (hunk 4 above)
Comment (rachitnigam) [inline js/src/test/scala/ParserTests.scala]: For the arguments, yes. For the function itself, no.

File: js/src/test/scala/BrowserTestSuite.scala (hunk 3 above)
Comment (rachitnigam) [inline js/src/test/scala/BrowserTestSuite.scala]: Yes, less char => Better code. That is why whitespace is the best programming language.

File: shared/src/main/scala/Syntax.scala (hunk 5 above)
Comment (rachitnigam) [inline shared/src/main/scala/Syntax.scala]: Write a combinator that setPos inside withPos?

**Comment by arjunguha** [comment]
Solid.

**Pull request merged**